### PR TITLE
fix(Scripts/BlackMorass): Fixed mob waves sequence.

### LINKED ISSUE #14849: [Opening the Dark Portal] Add spawning order is incorrect and no rep rewards for defeating waves.
### Current Behaviour

All 5 adds:
Infinite Whelps
Infinite Assassin
Infinite Chronomancer
Infinite Executioner
Infinite Vanquisher
can spawn at any time when the event is started.

No reputation gains when a wave is defeated.

### Expected Blizzlike Behaviour

Up to the 6th rift, Assassin, Whelps, Chronomancer spawn in sequence, repeating.
On the 7th rift and beyond, Executioner, Chronomancer, Whelps, Assassin spawn in sequence, repeating.
On the 13th rift and beyond, Executioner, Vanquisher, Chronomancer, Assassin spawn in sequence, repeating.

Once a wave is defeated, then the player receives 2 Keepers of Time (Faction 989) rep in Normal and 3 in Heroic.
Rep gains stop at friendly for defeating waves.

When the rift first spawns, there is a 15s timer for every single add spawned, including the elite.

### Source

Retail

### Steps to reproduce the problem

.go c id 15608
Observe which adds are spawning from which rifts

### Extra Notes

The only server message I found for the reputation gains was SMSG_SET_FACTION_STANDING, nothing else.

### AC rev. hash/commit

https://github.com/azerothcore/azerothcore-wotlk/commit/c3dd1b7a5ca0a9945d609213fdd3c4a19e0106e2

### Operating system

Windows 10

### Custom changes or Modules

_No response_

### PR DESCRIPTION
**Pull request by UltraNix**
Added reputation upon killing the wave.
Fixes #14849

## Issues Addressed:

- Closes #14849

## Tests Performed:

- Tested ingame.

## How to Test the Changes:

`.go c id 15608`
Observe which adds are spawning from which rifts

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by Gultask** [comment]
I thought that maybe the reputation gain comes from creature entries. I haven't pinned it down yet but there are several IDs, at least 2 for most adds. At first it seems random which one is used so I haven't investigated further as I still have plenty to check.

**Comment by Gultask** [comment]
Hey, Nix. First of all I need to apologise as I didn't imagine you would pick it up right away. And I left out some research I was planning to do. So there's incomplete data for this one.

First I can confirm that the rep gains are solely from Infinite Assassin (17835)

Second, every Add except the Whelps has another ID they can spawn in. For example, assassin can be either 21137 and 17835. The only difference between them seems to be that one gives rep (17835) and the other does not.
I've looked at this for about an hour now and I've seen no patter to their spawns.
It seems like either IDs can spawn at random. So the first 6 rifts will be:

1. Infinite Assassin (17835 OR 21137)
2. Infinite Whelp (21818)
3. Infinite Chronomancer (17892 OR 21136)

All IDs are here:
"17835"	"Infinite Assassin"
"21137"	"Infinite Assassin"
"17892"	"Infinite Chronomancer"
"21136"	"Infinite Chronomancer"
"18994"	"Infinite Executioner"
"21138"	"Infinite Executioner"
"18995"	"Infinite Vanquisher"
"21139"	"Infinite Vanquisher"
"21818"	"Infinite Whelp"

**Comment by UltraNix** [comment]
This sounds weird...
https://www.wowhead.com/tbc/faction=989/keepers-of-time#comments:id=50412
`15x normal min-boss mobs which reward 50 rep, 3x bosses which reward 120 rep`
So the Rift keepers give as much rep as trash?

**Comment by Gultask** [comment]
Only 2 and 3 from one Id for Normal and Heroic respectively from the adds.

In AC I noticed some heroic adds were giving rep when they shouldn't, though. Like 50 or so.

**Comment by UltraNix** [comment]
`A clear of Heroic Black Morass will give approximately 1725 rep (versus 1110 from the nonheroic version).`
That wound mean trash mobs should not give any rep.

**Comment by UltraNix** [comment]
> Only 2 and 3 from one Id for Normal and Heroic respectively from the adds.
> 
> In AC I noticed some heroic adds were giving rep when they shouldn't, though. Like 50 or so.

Yeah, but e.g. on wowhead both version of Infinite Chronomancer gives 50 on normal.

**Comment by Gultask** [comment]
Since it's only 1 ID giving rep in the single digits out of 8 possible IDs (around 50% of all waves) I wouldn't imagine it's a convenient way to farm rep over full clears so it makes sense people would leave out this info from guides.

**Comment by Gultask** [comment]
> > Only 2 and 3 from one Id for Normal and Heroic respectively from the adds.
> > 
> > In AC I noticed some heroic adds were giving rep when they shouldn't, though. Like 50 or so.
> 
> Yeah, but e.g. on wowhead both version of Infinite Chronomancer gives 50 on normal.

NoNo that'd be insane. Probably where the this bad data came from in the first place

**Comment by UltraNix** [comment]
@Gultask Updated it - please retest it.

**Comment by Gultask** [comment]
Only 17835 and its heroic version have rep rewards on kill.

**Comment by UltraNix** [comment]
No. Every mob have rep version.

**Comment by Gultask** [comment]
Where did you hear that from?

**Comment by UltraNix** [comment]
I mean the entries were already in database.

**Comment by Gultask** [comment]
I mentioned it before, I think.
Those entries are wrong and need to be deleted.

**Comment by UltraNix** [comment]
Done.

File: data/sql/updates/pending_db_world/rev_1675618471961745900.sql [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+(21137,0,6,0,25,0,100,0,0,0,0,0,0,11,31326,0,0,0,0,0,1,0,0,0,0,0,0,0,0,'Infinite Assassin - On Reset - Cast Spell Corrupt Medivh');
+
+DELETE FROM `creature_text` WHERE `CreatureID`=21137;
+INSERT INTO `creature_text` (`CreatureID`, `GroupID`, `ID`, `Text`, `Type`, `Language`, `Probability`, `Emote`, `Duration`, `Sound`, `BroadcastTextId`, `TextRange`, `comment`) VALUES
+(21137,0,0,'More will take my place.',12,0,100,0,0,0,15167,0,'Infinite Assassin'),
+(21137,0,1,'Our time has come!',12,0,100,0,0,0,15163,0,'Infinite Assassin'),
+(21137,0,2,'The wizard will fall!',12,0,100,0,0,0,15164,0,'Infinite Assassin');
+
+UPDATE `creature_onkill_reputation` SET `RewOnKillRepValue1`=2 WHERE `creature_id`=21137;
+UPDATE `creature_onkill_reputation` SET `RewOnKillRepValue1`=3 WHERE `creature_id`=22164;
```
Comment (Gultask) [inline data/sql/updates/pending_db_world/rev_1675618471961745900.sql:35]: This one as well

File: data/sql/updates/pending_db_world/rev_1675618471961745900.sql [hunk 2]
```diff
@@ -0,0 +1,124 @@
+--
+-- Infinite Assasin
+UPDATE `creature_template` SET `difficulty_entry_1`=20740 WHERE `entry`=17835;
+UPDATE `creature_template` SET `faction`=1720, `MovementType`=1 WHERE `entry`=20740;
+
+UPDATE `creature_template` SET `difficulty_entry_1`=22164, `speed_run`=1.78571, `AiName`='SmartAI' WHERE `entry`=21137;
```
Comment (Gultask) [inline data/sql/updates/pending_db_world/rev_1675618471961745900.sql:6]: Don't change speed values. They are sniffed.
https://github.com/azerothcore/azerothcore-wotlk/pull/14495

**Review by Gultask — CHANGES_REQUESTED** [review]

File: data/sql/updates/pending_db_world/rev_1675618471961745900.sql (hunk 1 above)
Comment (Gultask) [reply data/sql/updates/pending_db_world/rev_1675618471961745900.sql:35]: They are Assassins as well but like I said before only 17835 rewards rep

Comment (UltraNix) [reply data/sql/updates/pending_db_world/rev_1675618471961745900.sql:35]: These Assasins are the rep versions. Not every Assasin gives rep.

File: data/sql/updates/pending_db_world/rev_1675618471961745900.sql [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+INSERT INTO `creature_text` (`CreatureID`, `GroupID`, `ID`, `Text`, `Type`, `Language`, `Probability`, `Emote`, `Duration`, `Sound`, `BroadcastTextId`, `TextRange`, `comment`) VALUES
+(21139,0,0,'We are not finished!',12,0,100,0,0,0,15174,0,'Infinite Vanquisher'),
+(21139,0,1,'Death to the Last Guardian!',12,0,100,0,0,0,15171,0,'Infinite Vanquisher'),
+(21139,0,2,'We will not fail!',12,0,100,0,0,0,23332,0,'Infinite Vanquisher');
+
+DELETE FROM `creature_onkill_reputation` WHERE `creature_id` IN (21139,22168);
+
+-- Conditions
+DELETE FROM `conditions` WHERE `SourceTypeOrReferenceId`=22 AND `SourceEntry` IN (17835,17892,18994,18995,21136,21137,21138,21139);
+INSERT INTO`conditions` (`SourceTypeOrReferenceId`, `SourceGroup`, `SourceEntry`, `SourceId`, `ElseGroup`, `ConditionTypeOrReference`, `ConditionTarget`, `ConditionValue1`, `ConditionValue2`, `ConditionValue3`, `NegativeCondition`, `ErrorType`, `ErrorTextId`, `ScriptName`, `Comment`) values
```
Comment (Gultask) [inline data/sql/updates/pending_db_world/rev_1675618471961745900.sql:115]:
```suggestion
INSERT INTO`conditions` (`SourceTypeOrReferenceId`, `SourceGroup`, `SourceEntry`, `SourceId`, `ElseGroup`, `ConditionTypeOrReference`, `ConditionTarget`, `ConditionValue1`, `ConditionValue2`, `ConditionValue3`, `NegativeCondition`, `ErrorType`, `ErrorTextId`, `ScriptName`, `Comment`) VALUES
```

File: src/server/scripts/Kalimdor/CavernsOfTime/TheBlackMorass/instance_the_black_morass.cpp [hunk 4]
```diff
@@ -111,10 +111,14 @@ class instance_the_black_morass : public InstanceMapScript
                 case NPC_RIFT_LORD:
                 case NPC_RIFT_LORD_2:
                 case NPC_INFINITE_ASSASIN:
+                case NPC_INFINITE_ASSASIN_2:
                 case NPC_INFINITE_WHELP:
                 case NPC_INFINITE_CRONOMANCER:
+                case NPC_INFINITE_CRONOMANCER_2:
```
Comment (Gultask) [inline src/server/scripts/Kalimdor/CavernsOfTime/TheBlackMorass/instance_the_black_morass.cpp:117]: CHRONOMANCER

File: src/server/scripts/Kalimdor/CavernsOfTime/TheBlackMorass/the_black_morass.cpp [hunk 5]
```diff
@@ -64,9 +64,15 @@ enum medivhMisc
     EVENT_OUTRO_5               = 14,
     EVENT_OUTRO_6               = 15,
     EVENT_OUTRO_7               = 16,
-    EVENT_OUTRO_8               = 17
+    EVENT_OUTRO_8               = 17,
+
+    FACTION_KEEPERS_OF_TIME     = 989
```
Comment (Gultask) [inline src/server/scripts/Kalimdor/CavernsOfTime/TheBlackMorass/the_black_morass.cpp:69]: Is this still used? Or not needed anymore?
`FACTION_KEEPERS_OF_TIME     = 989`?

**Comment by Maelthyrr** [comment]
@Nyeriah is this pr still necessary since your last PR?

**Comment by Gultask** [comment]
Nyeriah's PR doesn't address the issues here, only some outstanding problems that were breaking the instance.
I'd wait for merge conflicts to be handled before you test it, though

**Comment by UltraNix** [comment]
Fixed.

**Review by Nyeriah — APPROVED** [review]

**Pull request merged**